hazelcast-client/src/test/java/com/hazelcast/client/heartbeat/ClientHeartbeatTest.java

```diff
@@ -0,0 +1,147 @@
+        hazelcastFactory.terminateAll();
+    }
+
+    @Test
+    public void testHeartbeatStoppedEvent() throws InterruptedException {
+        HazelcastInstance instance = hazelcastFactory.newHazelcastInstance();
+        ClientConfig clientConfig = new ClientConfig();
+        clientConfig.setProperty(ClientProperty.HEARTBEAT_TIMEOUT, "3000");
+        clientConfig.setProperty(ClientProperty.HEARTBEAT_INTERVAL, "500");
+        HazelcastInstance client = hazelcastFactory.newHazelcastClient(clientConfig);
+
+        HazelcastClientInstanceImpl clientImpl = getHazelcastClientInstanceImpl(client);
+        ClientConnectionManager connectionManager = clientImpl.getConnectionManager();
+        final CountDownLatch countDownLatch = new CountDownLatch(1);
+        connectionManager.addConnectionHeartbeatListener(new ConnectionHeartbeatListener() {
+            @Override
+            public void heartBeatStarted(Connection connection) {
+            }
+
+            @Override
+            public void heartBeatStopped(Connection connection) {
+                countDownLatch.countDown();
+            }
+        });
+
+        blockMessagesFromInstance(instance, client);
+        assertOpenEventually(countDownLatch);
+    }
+
+    @Test(expected = TargetDisconnectedException.class)
+    public void testInvocation_whenHeartbeatStopped() throws InterruptedException {
+        hazelcastFactory.newHazelcastInstance();
+        ClientConfig clientConfig = new ClientConfig();
+        clientConfig.setProperty(ClientProperty.HEARTBEAT_TIMEOUT, "3000");
+        clientConfig.setProperty(ClientProperty.HEARTBEAT_INTERVAL, "500");
+        HazelcastInstance client = hazelcastFactory.newHazelcastClient(clientConfig);
+        HazelcastInstance instance2 = hazelcastFactory.newHazelcastInstance();
+
+        String keyOwnedByInstance2 = generateKeyOwnedBy(instance2);
+        IMap map = client.getMap(randomString());
+        map.put(keyOwnedByInstance2, randomString());
+        blockMessagesFromInstance(instance2, client);
+        map.put(keyOwnedByInstance2, randomString());
+    }
+
+    @Test
+    public void testAsyncInvocation_whenHeartbeatStopped() throws InterruptedException {
+        hazelcastFactory.newHazelcastInstance();
+        ClientConfig clientConfig = new ClientConfig();
+        clientConfig.setProperty(ClientProperty.HEARTBEAT_TIMEOUT, "3000");
+        clientConfig.setProperty(ClientProperty.HEARTBEAT_INTERVAL, "500");
+        HazelcastInstance client = hazelcastFactory.newHazelcastClient(clientConfig);
+        HazelcastInstance instance2 = hazelcastFactory.newHazelcastInstance();
+
+        IMap map = client.getMap(randomString());
+        String keyOwnedByInstance2 = generateKeyOwnedBy(instance2);
+        map.put(keyOwnedByInstance2, randomString());
+        blockMessagesFromInstance(instance2, client);
+        ClientDelegatingFuture future = (ClientDelegatingFuture) map.putAsync(keyOwnedByInstance2, randomString());
```

Comment:
Don't you need to do async call before blocking the connection? Don't you need to simulate that the request arrived at the server but the response did not arrive due to connection.